org.eclipse.scout.rt.client/src/main/java/org/eclipse/scout/rt/client/ui/desktop/hybrid/uicallback/UiCallbacks.java

```diff
@@ -0,0 +1,480 @@
+   * @return The result as processed by the {@link IUiCallbackHandler}.
+   */
+  public <R> Future<R> send(IWidget owner, IUiCallbackHandler<?, R> handler) {
+    return send(owner, handler, null);
+  }
+
+  /**
+   * Calls the UI handler with ObjectType as returned by {@link IUiCallbackHandler#uiCallbackHandlerObjectType()} and
+   * returns its result as processed by the given {@link IUiCallbackHandler}.
+   *
+   * @param owner
+   *          The {@link IWidget} this request belongs to. Pass the {@link IWidget} that specifies the lifetime of this
+   *          UI callback. E.g. if the callback should be performed as long as the user is logged in, use
+   *          {@code IDesktop.CURRENT.get()}. The callback will automatically be cancelled as soon as this
+   *          {@link IWidget} is disposed.
+   * @param handler
+   *          An {@link IUiCallbackHandler} implementation that specifies the UI handler ObjectType and the processing
+   *          of the response.
+   * @param dataToSendToUi
+   *          An optional {@link IDoEntity} to send to the UI handler on the browser.
+   * @return The result as processed by the {@link IUiCallbackHandler}.
+   */
+  public <R> Future<R> send(IWidget owner, IUiCallbackHandler<?, R> handler, IDoEntity dataToSendToUi) {
+    return send(owner, handler, null, null);
+  }
+
+  /**
+   * Calls the UI handler with ObjectType as returned by {@link IUiCallbackHandler#uiCallbackHandlerObjectType()} and
+   * returns its result as processed by the given {@link IUiCallbackHandler}.
+   *
+   * @param owner
+   *          The {@link IWidget} this request belongs to. Pass the {@link IWidget} that specifies the lifetime of this
+   *          UI callback. E.g. if the callback should be performed as long as the user is logged in, use
+   *          {@code IDesktop.CURRENT.get()}. The callback will automatically be cancelled as soon as this
+   *          {@link IWidget} is disposed.
+   * @param handler
+   *          An {@link IUiCallbackHandler} implementation that specifies the UI handler ObjectType and the processing
+   *          of the response.
+   * @param dataToSendToUi
+   *          An optional {@link IDoEntity} to send to the UI handler on the browser.
+   * @param callbackId
+   *          An optional callbackId. All callbacks with the same ID are only sent once to the UI and on a response all
+   *          pending callbacks with this ID are resolved at once. By default, each callback has its own unique ID.
+   * @return The result as processed by the {@link IUiCallbackHandler}.
+   */
+  public <R> Future<R> send(IWidget owner, IUiCallbackHandler<?, R> handler, IDoEntity dataToSendToUi, String callbackId) {
+    assertNotNull(handler);
+    assertNotNull(owner);
+    assertModelThread();
+
+    if (StringUtility.isNullOrEmpty(callbackId)) {
+      callbackId = UUID.randomUUID().toString();
+    }
+
+    boolean isNewId;
+    P_UiCallback<R> callback;
+    synchronized (m_pendingUiCallbacks) {
```

Comment:
Why does it need to be synchronized? Isn't it always called from a model job?

Reply:
So that the future can be used outside a ModelThread (e.g. when calling cancel)